python/dgl/backend/pytorch/sparse_emb.py

```diff
@@ -0,0 +1,418 @@
+import torch as th
+import abc
+from abc import abstractmethod
+from .tensor import attach_grad, is_recording
+
+class GraphSparseEmbedding:
+    '''Sparse embeddings for graph.
+
+    DGL provides a sparse embedding to support models that require learnable embeddings.
+    DGL's sparse embeddings are mainly used for learning node embeddings of graph models.
+    The sparse embeddings have to be updated by DGL's optimizers instead of
+    the optimizers provided by the deep learning frameworks (e.g., Pytorch).
+
+    To support efficient training on a graph with many nodes, the embeddings support sparse
+    updates. That is, only the embeddings involved in a mini-batch computation are updated.
+    Currently, DGL provides two optimizer: `SparseAdagrad` and `SparseAdam. DGL will provide more
+    optimizers in the future.
+
+    Parameters
+    ----------
+    num_embeddings : int
+        The number of embeddings. Currently, the number of embeddings has to be the same as
+        the number of nodes or the number of edges.
+    embedding_dim : int
+        The dimension size of embeddings.
+    device : th.device
+        Target device to manipulate embedding related opations
+    name : str
+        The name of the embeddings. The name should uniquely identify embeddings in a system.
+    init_func : callable, optional
+        The function to create the initial data. If the init function is not provided,
+        the values of the embeddings are initialized to zero.
+    rank : int, optional
+        The rank in multi-gpu process group. -1 means single GPU or CPU training. Default: -1.
+    world_size : int, optional
+        Number of devices used in mutli-gpu training. Default: 0.
+    queues : list of multiprocessing.Queue, optional
+        Used by multi-gpu processes to share embedding tensor. Default: None.
```

Comment:
The main difference between this API and the one in DistDGL is that DistDGL knows all information of the "world" when it is launched. You actually can get this information from Pytorch.

Reply:
Which API provides the world_size?

Reply:
world size is the number of clients. This can be calculated given the number of machines and the number of trainers per machine.